[Serialization] Refactor normal conformance serialization

xedin opened the pull request:
Make it so requirement substitutions are serialized/deserialized independently from
synthetic context (because these two are not always related), also optimize storage
of the witness to avoid unnecessary allocation and remove `requiresSubstitution`
invariant which is not always correctly established and not really needed.

Resolves [SR-6754](https://bugs.swift.org/browse/SR-6754).
Resolves: rdar://problem/36497404

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

slavapestov commented on validation-test/compiler_crashers_2_fixed/0153-rdar36497404.swift:
```diff
@@ -0,0 +1,44 @@
+// RUN: %target-build-swift -emit-module -o %t %s
```
It should be possible to produce a smaller test case, and it should go in test/Serialization

slavapestov replied:
It should still be in test/Serialization since its testing serialization

xedin replied:
I've put it into crashers because it was a crash, but I can put it where ever you tell me to put it.

slavapestov commented on include/swift/AST/Witness.h:
```diff
@@ -168,6 +168,11 @@ class Witness {
   }
 
+  /// Determine if the witness may have requirement substitutions.
+  bool mayHaveRequirementSubstitutions() const {
```
Does getSyntheticEnvironment not imply requiresSubstitution?

slavapestov replied:
What's the specific case where requiresSubstitution() is true and getSyntheticEnvironment() is false? What about vice versa? I think it would help to understand what's actually going on here.

xedin replied:
As far as I understand `requiresSubstitution` covers both requirement and witness substitutions but the latter doesn't require generic env to be present, what is happening in the example I have - `a` would not have generic environment but still requires substitutions because it's a static witness (I think). There is a single requirement substitution in the map from S to S (I think because it's static witness) but information about generic env and number of requirement substitutions is not going to be stored altough that single requirement substitution still would which offsets every other substitution by one.

xedin replied:
I might not be understanding completely what is going on here but what you suggest doesn't actually fix the problem (I tried moving `if` block from deserialization to Witness general constructor) ... Because if you consider declaration of `a` in `P3` it does return `true` for `requiresSubstitution`, doesn't have generic synthetic environment (which seems correct), does have a single synthetic requirement substitution `S`, and one witness substitution `S` when it's being originally written to partial module so checking just generic signature at one place and `requiresSubstitution()` when writing records would produce different results in serialization that's why deserialization produces incorrect results.

It looks like https://github.com/apple/swift/blob/master/lib/Sema/TypeCheckProtocol.cpp#L93-L101 is getting synthetic substitutions independent from presence of synthetic environment, that's why there could be requirement substitutions without synthetic environment, maybe the bug hides in that fact?

DougGregor replied:
I think there are several bugs here, one of which is tripped up by the test case you've reduced. Here are the issues I see:

1. `requiresSubstitution()` is checking the kind of storage (`StoredWitness *`) as a proxy for "has interesting an synthetic environment", but nothing establishes that as an invariant. We can either establish that as an invariant or eliminate `requiresSubstitution()` outright. It's probably best to remove it---it's used incorrectly/inconsistently basically everywhere, mainly to dodge the assertions in `getSyntheticEnvironment()` and `getRequirementToSyntheticSubs()`. Let's make those functions do a `dyn_cast` to `StoredWitness *` and return its information when available, or nullptr/{} when not available, so we can clean up the callers. This is what my first big comment was about, but it's not the whole solution.

2. You've found that one of the assumptions that serialization is making is not an invariant in the AST, i.e., that one can only have a non-empty `reqToSyntheticSubs` if the requirement environment is non-null. You pull request fixes this by using a consistent predicate in both places, which is good, but...

3.  When this happens, serialization (both before and after your PR) will drop the contents of `reqToSyntheticSubs`, so the deserialized `Witness` will be different from the originally-constructed form. This could potentially break [SILGen's protocol witness emission](https://github.com/apple/swift/blob/master/lib/SILGen/SILGenType.cpp#L607) from a deserialized `Witness`, which looks at and uses `reqToSyntheticSubs` always (without checking for a null requirement environment). To fix this, I recommend we always emit the size of `reqToSyntheticSubs` (e.g., [take out this `if` on the serialization side](https://github.com/apple/swift/blob/master/lib/Serialization/Serialization.cpp#L1541) and the [corresponding one in deserialization](https://github.com/apple/swift/blob/master/lib/Serialization/Deserialization.cpp#L5313-L5318)).

xedin replied:
Thanks @DougGregor! What you are saying totally makes sense to me, I don't think number of requirement substitution should be dependent on presence of synthetic environment!

slavapestov reviewed (commented):
The commit message is not very clear. It should explain what the two conditions mean in terms of language features used and why the old one was insufficient. Right now you’re just paraphrasing the jargon in the diff.

xedin committed:
[Serialization] Always serialize requirement substitutions

Absence of synthetic generic environment should not affect
serialization of the required substitutions because they can
come from outer requirement context for static members.

Resolves: rdar://problem/36497404

xedin committed:
[AST] Unify witness construction logic

Avoid storing more information than necessary in witnesses
and establishing the invariant that we only use out-of-line
storage when we require substitutions.

xedin committed:
[AST] Remove `requiresSubstitution` invariant from `Witness`

xedin committed:
[Serialization] Update Module Format version

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

DougGregor reviewed (approved):
This looks fantastic, thank you!

slavapestov commented:
@xedin Thank you for taking the time to get to the bottom of the problem and come up with this great fix!
